### PR TITLE
`<generator>`: An attempt to merge `_Top` and `_Info`

### PR DESCRIPTION
**Pull request by JMazurkiewicz**
Merge `promise_type::_Top` and `promise_type::_Info` into single `promise_type::_Data` member of type `uintptr_t`.

### CODE REVIEW

**Comment by JMazurkiewicz** [comment]
> For my own education, I'd like to confirm - the size of `promise_type` is important enough to be worth this additional complexity?

It's debatable, that's why I called this PR `An attempt`.

My opinion is yes - it is important. Since promise object is dynamically allocated, smaller size can be useful when users supply their own allocator and there is not a lot of memory to give. At first it seems that saving 4/8 bytes is not important, but when you think about recursive case, then we can gain much more. Also, the additional complexity is negligible in non-recursive case - for example, a call to `_Get_top` should be optimized away in `/O1` build (I'm going to verify it).

Maybe we should mark this PR as `decision needed`?

**Comment by StephanTLavavej** [comment]
Being dynamically allocated is compelling enough to me, thanks!

File: stl/inc/generator [hunk 1]
```diff
@@ -365,13 +363,35 @@ private:
...
+    void _Set_top(coroutine_handle<_Base> _Top) noexcept {
+        _Data = reinterpret_cast<uintptr_t>(_Top.address());
+    }
+
     template <class, class>
     friend struct _Gen_iter_provider;
 
-    // _Top and _Info are mutually exclusive, and could potentially be merged.
-    coroutine_handle<_Base> _Top = coroutine_handle<_Base>::from_promise(*this);
+    // First bit of `_Data` indicates stored information:
```
[the remark this answers is not in the document]
Comment (JMazurkiewicz) [reply stl/inc/generator:390]: I agree, this should be changed later. For now, it's not worth resetting CI.

File: stl/inc/generator [hunk 2]
```diff
@@ -365,13 +363,35 @@ private:
...
+    // 0 - `_Data` is a top coroutine handle,
+    // 1 - `_Data ^ 1U` is a pointer to an object of type `_Nest_info`.
```
Comment (StephanTLavavej) [inline stl/inc/generator:392]: Really nitpicky, but these dashes kind of look like minus operators during a quick read.

File: stl/inc/generator [hunk 3]
```diff
@@ -311,14 +311,13 @@ private:
             _STL_INTERNAL_STATIC_ASSERT(is_pointer_interconvertible_base_of_v<_Base, _CoroPromise>);
 #endif // ^^^ no workaround ^^^
 
-            _Base& _Current = _Handle.promise();
-            if (!_Current._Info) {
+            _Nest_info* _Info = _Handle.promise()._Try_get_nest_info();
+            if (!_Info) {
                 return _STD noop_coroutine();
             }
```
Comment (StephanTLavavej) [inline stl/inc/generator:317]: Style suggestion: This is a minimal transformation from the existing code (a good thing, considered in isolation), but I think a slightly larger change would be even better. The other uses are scoping `_Info` to an `if`-statement, and that would seem to be natural here too.

File: stl/inc/generator [hunk 4]
```diff
@@ -266,7 +266,7 @@ public:
     void return_void() const noexcept {}
 
     void unhandled_exception() {
-        if (_Info) {
+        if (auto _Info = _Try_get_nest_info()) {
```
Comment (StephanTLavavej) [inline stl/inc/generator:269]: We could make this `const auto`, and do that consistently below.

File: stl/inc/generator [hunk 5]
```diff
@@ -311,14 +311,13 @@ private:
...
-            _Base& _Current = _Handle.promise();
-            if (!_Current._Info) {
+            _Nest_info* _Info = _Handle.promise()._Try_get_nest_info();
+            if (!_Info) {
                 return _STD noop_coroutine();
             }
 
-            coroutine_handle<_Base> _Cont        = _Current._Info->_Parent;
-            _Current._Info->_Root.promise()._Top = _Cont;
-            _Current._Info                       = nullptr;
```
Comment (StephanTLavavej) [inline stl/inc/generator:321]: I see that `_Current._Info = nullptr;` (i.e. `_Handle.promise()._Info = nullptr;`) is being dropped, but I don't completely understand why.

The new `_Info->_Root.promise()._Set_top(_Cont);` (i.e. `_Handle.promise()._Try_get_nest_info()->_Root.promise()._Set_top(_Cont);`) seems to be modifying a different promise - are they actually the same?

(Note that I am very uninformed about the data structure relationships here.)

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/generator (hunk 5 above)
Comment (JMazurkiewicz) [reply stl/inc/generator:321]: > I see that `_Current._Info = nullptr;` (i.e. `_Handle.promise()._Info = nullptr;`) is being dropped, but I don't completely understand why.

We are in final awaiter, so current coroutine is about to be destroyed. Setting `_Current._Info` to `nullptr` is unnecessary.

> The new `_Info->_Root.promise()._Set_top(_Cont);` (i.e. `_Handle.promise()._Try_get_nest_info()->_Root.promise()._Set_top(_Cont);`) seems to be modifying a different promise - are they actually the same?

`_Info->_Root.promise()._Set_top(_Cont);` (used to be `_Current._Info->_Root.promise()._Top = _Cont;`) modifies root promise which in this situation cannot be the same as `_Handle.promise()`.

If `_Handle.promise()` was the same as root, then `_Handle.promise._Try_get_nest_info()` would be equal to `nullptr` (there is no nested generator), and in such case we return `noop_coroutine`.

Comment (StephanTLavavej) [reply stl/inc/generator:321]: Got it, thanks! :heart_eyes_cat:

**Pull request merged**